File: repos/firecrawl/apps/go-sdk/http_client.go
package firecrawl

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"mime/multipart"
	"net/http"
	"net/textproto"
	"strings"
	"time"
)

const (
	defaultAPIURL        = "https://api.firecrawl.dev"
	defaultTimeout       = 5 * time.Minute
	defaultMaxRetries    = 3
	defaultBackoffFactor = 0.5
)

// httpClient is the internal HTTP client for the Firecrawl API.
type httpClient struct {
	client        *http.Client
	apiKey        string
	baseURL       string
	maxRetries    int
	backoffFactor float64
	extraHeaders  map[string]string
}

func newHTTPClient(apiKey, baseURL string, client *http.Client, maxRetries int, backoffFactor float64, extraHeaders map[string]string) *httpClient {
	baseURL = strings.TrimRight(baseURL, "/")
	return &httpClient{
		client:        client,
		apiKey:        apiKey,
		baseURL:       baseURL,
		maxRetries:    maxRetries,
		backoffFactor: backoffFactor,
		extraHeaders:  extraHeaders,
	}
}

// post sends a POST request with a JSON body.
func (h *httpClient) post(ctx context.Context, path string, body interface{}, extraHeaders map[string]string) (json.RawMessage, error) {
	url := h.baseURL + path
	return h.doJSON(ctx, "POST", url, body, extraHeaders)
}

// patch sends a PATCH request.
func (h *httpClient) patch(ctx context.Context, path string, body interface{}) (json.RawMessage, error) {
	url := h.baseURL + path
	return h.doJSON(ctx, "PATCH", url, body, nil)
}

// get sends a GET request.
func (h *httpClient) get(ctx context.Context, path string) (json.RawMessage, error) {
	url := h.baseURL + path
	return h.doJSON(ctx, "GET", url, nil, nil)
}

// getAbsolute sends a GET request to an absolute URL (for pagination cursors).
func (h *httpClient) getAbsolute(ctx context.Context, absoluteURL string) (json.RawMessage, error) {
	return h.doJSON(ctx, "GET", absoluteURL, nil, nil)
}

// delete sends a DELETE request.
func (h *httpClient) delete(ctx context.Context, path string) (json.RawMessage, error) {
	url := h.baseURL + path
	return h.doJSON(ctx, "DELETE", url, nil, nil)
}

// postMultipart sends a POST request with a multipart/form-data body. The extra
// text `fields` are written first, followed by a single file part.
func (h *httpClient) postMultipart(
	ctx context.Context,
	path string,
	fields map[string]string,
	fileField, fileName, fileContentType string,
	fileContent []byte,
) (json.RawMessage, error) {
	url := h.baseURL + path

	buildBody := func() (io.Reader, string, error) {
		buf := &bytes.Buffer{}
		writer := multipart.NewWriter(buf)

		for k, v := range fields {
			if err := writer.WriteField(k, v); err != nil {
				return nil, "", err
			}
		}

		partHeader := make(textproto.MIMEHeader)
		partHeader.Set(
			"Content-Disposition",
			fmt.Sprintf(`form-data; name=%q; filename=%q`, fileField, fileName),
		)
		if fileContentType != "" {
			partHeader.Set("Content-Type", fileContentType)
		}
		part, err := writer.CreatePart(partHeader)
		if err != nil {
			return nil, "", err
		}
		if _, err := part.Write(fileContent); err != nil {
			return nil, "", err
		}
		if err := writer.Close(); err != nil {
			return nil, "", err
		}
		return buf, writer.FormDataContentType(), nil
	}

	body, contentType, err := buildBody()
	if err != nil {
		return nil, &FirecrawlError{Message: fmt.Sprintf("failed to build multipart body: %v", err)}
	}

	var lastErr error
	for attempt := 0; attempt <= h.maxRetries; attempt++ {
		if attempt > 0 {
			if err := h.sleepBackoff(ctx, attempt); err != nil {
				return nil, err
			}
			body, contentType, err = buildBody()
			if err != nil {
				return nil, &FirecrawlError{Message: fmt.Sprintf("failed to rebuild multipart body: %v", err)}
			}
		}

		req, err := http.NewRequestWithContext(ctx, "POST", url, body)
		if err != nil {
			return nil, &FirecrawlError{Message: fmt.Sprintf("failed to create request: %v", err)}
		}

		if h.apiKey != "" {
			req.Header.Set("Authorization", "Bearer "+h.apiKey)
		}
		req.Header.Set("Content-Type", contentType)
		req.Header.Set("User-Agent", "firecrawl-go/"+Version)
		for k, v := range h.extraHeaders {
			req.Header.Set(k, v)
		}

		resp, err := h.client.Do(req)
		if err != nil {
			if ctx.Err() != nil {
				return nil, ctx.Err()
			}
			lastErr = err
			continue
		}

		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			lastErr = err
			continue
		}

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			return json.RawMessage(respBody), nil
		}

		errMsg, errCode := extractError(respBody, resp.StatusCode)

		switch resp.StatusCode {
		case 401:
			return nil, &AuthenticationError{
				FirecrawlError: FirecrawlError{StatusCode: 401, ErrorCode: errCode, Message: errMsg},
			}
		case 429:
			return nil, &RateLimitError{
				FirecrawlError: FirecrawlError{StatusCode: 429, ErrorCode: errCode, Message: errMsg},
			}
		}

		if resp.StatusCode >= 400 && resp.StatusCode < 500 && resp.StatusCode != 408 && resp.StatusCode != 409 {
			return nil, &FirecrawlError{StatusCode: resp.StatusCode, ErrorCode: errCode, Message: errMsg}
		}

		lastErr = &FirecrawlError{StatusCode: resp.StatusCode, ErrorCode: errCode, Message: errMsg}
	}

	if lastErr != nil {
		if fe, ok := lastErr.(*FirecrawlError); ok {
			return nil, fe
		}
		return nil, &FirecrawlError{Message: fmt.Sprintf("request failed after %d retries: %v", h.maxRetries, lastErr)}
	}
	return nil, &FirecrawlError{Message: "request failed"}
}

func (h *httpClient) doJSON(ctx context.Context, method, url string, body interface{}, extraHeaders map[string]string) (json.RawMessage, error) {
	var bodyReader io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return nil, &FirecrawlError{Message: fmt.Sprintf("failed to serialize request body: %v", err)}
		}
		bodyReader = bytes.NewReader(data)
	}

	var lastErr error
	for attempt := 0; attempt <= h.maxRetries; attempt++ {
		if attempt > 0 {
			if err := h.sleepBackoff(ctx, attempt); err != nil {
				return nil, err
			}

			// Reset the body reader for retries.
			if body != nil {
				data, _ := json.Marshal(body)
				bodyReader = bytes.NewReader(data)
			}
		}

		req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
		if err != nil {
			return nil, &FirecrawlError{Message: fmt.Sprintf("failed to create request: %v", err)}
		}

		if h.apiKey != "" {
			req.Header.Set("Authorization", "Bearer "+h.apiKey)
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", "firecrawl-go/"+Version)
		// Apply client-level headers.
		for k, v := range h.extraHeaders {
			req.Header.Set(k, v)
		}
		// Apply per-request headers (override client-level).
		for k, v := range extraHeaders {
			req.Header.Set(k, v)
		}

		resp, err := h.client.Do(req)
		if err != nil {
			// If context is cancelled, return immediately instead of retrying.
			if ctx.Err() != nil {
				return nil, ctx.Err()
			}
			lastErr = err
			continue // Retry on transport errors.
		}

		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			lastErr = err
			continue
		}

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			return json.RawMessage(respBody), nil
		}

		// Parse error details from the response.
		errMsg, errCode := extractError(respBody, resp.StatusCode)

		// Non-retryable client errors.
		switch resp.StatusCode {
		case 401:
			return nil, &AuthenticationError{
				FirecrawlError: FirecrawlError{StatusCode: 401, ErrorCode: errCode, Message: errMsg},
			}
		case 429:
			return nil, &RateLimitError{
				FirecrawlError: FirecrawlError{StatusCode: 429, ErrorCode: errCode, Message: errMsg},
			}
		}

		if resp.StatusCode >= 400 && resp.StatusCode < 500 && resp.StatusCode != 408 && resp.StatusCode != 409 {
			return nil, &FirecrawlError{StatusCode: resp.StatusCode, ErrorCode: errCode, Message: errMsg}
		}

		// Retryable: 408, 409, 5xx
		lastErr = &FirecrawlError{StatusCode: resp.StatusCode, ErrorCode: errCode, Message: errMsg}
	}

	if lastErr != nil {
		if fe, ok := lastErr.(*FirecrawlError); ok {
			return nil, fe
		}
		return nil, &FirecrawlError{Message: fmt.Sprintf("request failed after %d retries: %v", h.maxRetries, lastErr)}
	}
	return nil, &FirecrawlError{Message: "request failed"}
}

func (h *httpClient) sleepBackoff(ctx context.Context, attempt int) error {
	delay := time.Duration(h.backoffFactor*1000*math.Pow(2, float64(attempt-1))) * time.Millisecond
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(delay):
		return nil
	}
}

// extractError parses an API error response to get the message and error code.
func extractError(body []byte, statusCode int) (string, string) {
	var parsed map[string]interface{}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return fmt.Sprintf("HTTP %d error", statusCode), ""
	}

	msg := fmt.Sprintf("HTTP %d error", statusCode)
	if v, ok := parsed["error"]; ok {
		msg = fmt.Sprintf("%v", v)
	} else if v, ok := parsed["message"]; ok {
		msg = fmt.Sprintf("%v", v)
	}

	var errCode string
	if v, ok := parsed["code"]; ok && v != nil {
		errCode = fmt.Sprintf("%v", v)
	}

	return msg, errCode
}

File: repos/firecrawl/apps/ruby-sdk/lib/firecrawl/http_client.rb
# frozen_string_literal: true

require "net/http"
require "json"
require "securerandom"
require "uri"

module Firecrawl
  # Internal HTTP client for making authenticated requests to the Firecrawl API.
  # Handles retry logic with exponential backoff.
  #
  # @api private
  class HttpClient
    def initialize(api_key:, base_url:, timeout:, max_retries:, backoff_factor:)
      @api_key = api_key
      @base_url = base_url.chomp("/")
      @timeout = timeout
      @max_retries = max_retries
      @backoff_factor = backoff_factor
    end

    # Sends a POST request with JSON body.
    def post(path, body, extra_headers: {})
      uri = URI("#{@base_url}#{path}")
      request = Net::HTTP::Post.new(uri)
      request["Authorization"] = "Bearer #{@api_key}" if @api_key
      request["Content-Type"] = "application/json"
      extra_headers.each { |k, v| request[k] = v }
      request.body = JSON.generate(body)
      execute_with_retry(uri, request)
    end

    # Sends a GET request.
    def get(path)
      uri = URI("#{@base_url}#{path}")
      request = Net::HTTP::Get.new(uri)
      request["Authorization"] = "Bearer #{@api_key}" if @api_key
      execute_with_retry(uri, request)
    end

    # Sends a GET request with a full URL (for following next-page cursors).
    # Only sends the Authorization header if the URL matches the configured API origin.
    def get_absolute(absolute_url)
      uri = URI(absolute_url)
      base_uri = URI(@base_url)
      unless uri.host == base_uri.host && uri.port == base_uri.port && uri.scheme == base_uri.scheme
        raise FirecrawlError, "Absolute URL origin (#{uri.scheme}://#{uri.host}:#{uri.port}) does not match API base URL origin (#{base_uri.scheme}://#{base_uri.host}:#{base_uri.port}). Refusing to send credentials."
      end
      request = Net::HTTP::Get.new(uri)
      request["Authorization"] = "Bearer #{@api_key}" if @api_key
      execute_with_retry(uri, request)
    end

    # Sends a DELETE request.
    def delete(path)
      uri = URI("#{@base_url}#{path}")
      request = Net::HTTP::Delete.new(uri)
      request["Authorization"] = "Bearer #{@api_key}" if @api_key
      execute_with_retry(uri, request)
    end

    # Sends a PATCH request with JSON body.
    def patch(path, body)
      uri = URI("#{@base_url}#{path}")
      request = Net::HTTP::Patch.new(uri)
      request["Authorization"] = "Bearer #{@api_key}" if @api_key
      request["Content-Type"] = "application/json"
      request.body = JSON.generate(body)
      execute_with_retry(uri, request)
    end

    # Sends a POST request with a multipart/form-data body.
    #
    # @param path [String] API path
    # @param fields [Hash{String=>String}] additional form fields to include
    # @param file_field [String] form field name for the file part (e.g. "file")
    # @param filename [String] filename to send with the file part
    # @param content [String] raw bytes for the file part
    # @param content_type [String, nil] optional MIME type for the file part
    def post_multipart(path, fields:, file_field:, filename:, content:, content_type: nil)
      uri = URI("#{@base_url}#{path}")
      boundary = "----FirecrawlBoundary#{SecureRandom.hex(16)}"
      body = build_multipart_body(boundary, fields, file_field, filename, content, content_type)

      builder = lambda do
        request = Net::HTTP::Post.new(uri)
        request["Authorization"] = "Bearer #{@api_key}" if @api_key
        request["Content-Type"] = "multipart/form-data; boundary=#{boundary}"
        request.body = body
        request
      end

      execute_with_retry(uri, builder.call, request_builder: builder)
    end

    private

    def build_multipart_body(boundary, fields, file_field, filename, content, content_type)
      parts = +""
      fields.each do |name, value|
        parts << "--#{boundary}\r\n"
        parts << %(Content-Disposition: form-data; name="#{name}"\r\n\r\n)
        parts << value.to_s
        parts << "\r\n"
      end

      parts << "--#{boundary}\r\n"
      safe_file_field = file_field.to_s.gsub(/[\r\n"]/, "_")
      safe_filename = filename.to_s.gsub(/[\r\n"]/, "_")
      parts << %(Content-Disposition: form-data; name="#{safe_file_field}"; filename="#{safe_filename}"\r\n)
      parts << "Content-Type: #{content_type || "application/octet-stream"}\r\n\r\n"
      parts.force_encoding(Encoding::ASCII_8BIT)
      parts << content.to_s.dup.force_encoding(Encoding::ASCII_8BIT)
      parts << "\r\n--#{boundary}--\r\n"
      parts
    end

    def execute_with_retry(uri, request, request_builder: nil)
      attempt = 0
      loop do
        response = perform_request(uri, request)
        code = response.code.to_i
        body_str = response.body || ""

        if code >= 200 && code < 300
          return body_str.empty? ? {} : JSON.parse(body_str)
        end

        error_message = extract_error_message(body_str, code)
        error_code = extract_error_code(body_str)

        # Non-retryable client errors
        if code == 401
          raise AuthenticationError.new(error_message, error_code: error_code)
        end
        if code == 429
          raise RateLimitError.new(error_message, error_code: error_code)
        end
        if code >= 400 && code < 500 && code != 408 && code != 409
          raise FirecrawlError.new(error_message, status_code: code, error_code: error_code)
        end

        # Retryable errors: 408, 409, 502, 5xx
        if attempt < @max_retries
          attempt += 1
          sleep_with_backoff(attempt)
          request = request_builder.call if request_builder
          next
        end

        raise FirecrawlError.new(error_message, status_code: code, error_code: error_code)
      rescue Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::ETIMEDOUT,
             Net::OpenTimeout, Net::ReadTimeout, IOError => e
        if attempt < @max_retries
          attempt += 1
          sleep_with_backoff(attempt)
          request = request_builder.call if request_builder
          retry
        end
        raise FirecrawlError.new("Request failed: #{e.message}")
      end
    end

    def perform_request(uri, request)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")
      http.open_timeout = @timeout
      http.read_timeout = @timeout
      http.write_timeout = @timeout
      http.request(request)
    end

    def extract_error_message(body, status_code)
      parsed = JSON.parse(body)
      parsed["error"] || parsed["message"] || "HTTP #{status_code} error"
    rescue JSON::ParserError
      "HTTP #{status_code} error"
    end

    def extract_error_code(body)
      parsed = JSON.parse(body)
      code = parsed["code"]
      code&.to_s
    rescue JSON::ParserError
      nil
    end

    def sleep_with_backoff(attempt)
      delay = @backoff_factor * (2**(attempt - 1))
      sleep(delay)
    end
  end
end
